### PR TITLE
mgr: fix py3 support

### PR DESCRIPTION
**Pull request by tchaikov**
http://tracker.ceph.com/issues/22880

### CODE REVIEW

**Comment by rjfd** [comment]
@tchaikov shouldn't we also add the line below to all python files to make sure that all `import` statements work in both python 2 and 3?

`from __future__ import absolute_import`

**Comment by tchaikov** [comment]
@rjfd to make sure import statement works across py3 and py2, we need to add this line to all source files where we could have relative "import"s. but yes, it helps to prevent relative imoprts in future. i am adding another commit in this PR to add it to some source files i switched to relative import.

**Comment by jan--f** [comment]
I might be doing something wrong but I can't actually get vstart to run the mgr modules with python3. I pass ```-DWITH_PYTHON3=ON``` to cmake and python3 is the default on the build machine (though python2 is installed). The vstart logs however show:
```
 grep -i python out/*.log                                          :(
out/mgr.x.log:2018-02-09 12:31:54.236 7ff4bef70700  1 mgr[py] Loading python module 'balancer'
out/mgr.x.log:2018-02-09 12:31:54.244 7ff4bef70700 10 mgr[py] Computed sys.path '/home/jan/code/ceph/ceph_tchaikov/src/pybind:/home/jan/code/ceph/ceph_tchaikov/build/lib/cython_modules/lib.2::/usr/lib/python27.zip:/usr/lib64/python2.7:/usr/lib64/python2.7/plat-linux2:/usr/lib64/python2.7/lib-tk:/usr/lib64/python2.7/lib-old:/usr/lib64/python2.7/lib-dynload:/usr/lib64/python2.7/site-packages:/usr/lib64/site-python:/usr/lib/python2.7/site-packages:/usr/lib/site-python:/home/jan/code/ceph/ceph_tchaikov/src/pybind/mgr'
...
```
Is [this line](https://github.com/tchaikov/ceph/blob/7651c4ef6ecd0dafe0646f50ada722ad3373d4cb/CMakeLists.txt#L445) maybe a problem?

**Comment by smithfarm** [comment]
@jan--f Could you try `--DWITH_PYTHON3=ON --DMGR_PYTHON_VERSION=3` ?

**Comment by tchaikov** [comment]
@jan--f as @smithfarm pointed out, the `/home/jan/code/ceph/ceph_tchaikov/build/lib/cython_modules/lib.2` part is configured using `MGR_PYTHON_VERSION` option.

**Comment by jan--f** [comment]
@smithfarm @tchaikov Thx, I guess I could have figured that one out myself.

There are several calls to ```iteritems()``` which will cause an exception in python3. But I assume you're not trying to fix those here but rather leave that to the module authors?

File: src/pybind/mgr/restful/module.py [hunk 1]
```diff
@@ -251,7 +256,7 @@ def serve(self):
                 self._serve()
                 self.server.socket.close()
             except CannotServe as cs:
-                self.log.warn("server not running: {0}".format(cs.message))
+                self.log.warn("server not running: %s", cs)
```
Comment (jan--f) [inline src/pybind/mgr/restful/module.py]: I personally prefer the ```.format()``` formatting.

Comment (tchaikov) [reply src/pybind/mgr/restful/module.py]: i am not changing line for no reason: `Exception` in py3 does not have the `message` property. and i think using the first param as format string is well documented by `logging.debug()`, moreover, it's simpler and shorter.

Comment (jan--f) [reply src/pybind/mgr/restful/module.py]: ah yes...my mistake.

**Comment by tchaikov** [comment]
@rjfd it turns out that enforcing the relative/absolute import is a non-trivial task. for instance, in the restful module, we have circular import: module => hooks => module. it seems it's impossible to import a whole file intra-module, but that's the only way to get over the circular import. `from foo import bar` will give error like
```
ImportError: cannot import name instance
```

i am dropping 7651c4e for now, probably we can have it in another PR.

**Comment by rjfd** [comment]
@tchaikov ok, let's postpone that.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-10_17:19:49-rados-master-distro-basic-smithi/

**Comment by tchaikov** [comment]
changelog

- in addition to `Module`, as expose `StandbyModule` in `__init__.py`
- fix the py3 support in dashboard module
- fix the log message in `qa/tasks/mgr/test_dashboard.py`

**Comment by tchaikov** [comment]
@jcsp could you take a look? at the first glance, 410ce76 might be just cosmetic change, but it is not. without it, in my test bed, the `ceph_module` module can not be found by ceph-mgr modules if ceph-mgr is compiled with py3 support, i addressed this issue using `PyImport_AppendInittab()`.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-11_07:50:31-rados-wip-mgr-py3-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-02-09_16:51:44-rados-wip-kefu-testing-2018-02-09-2315-distro-basic-mira/

**Comment by jcsp** [comment]
The 410ce76 doesn't look problematic, but I don't understand the reason it was failing otherwise?  It should be legal to construct new modules in a subinterpreter and have .py modules see them :-/

**Comment by tchaikov** [comment]
hmmm, neither do i understand why. i was attempted to debug the python runtime. but cowardly stopped when i found that `PyImport_AppendInittab()` works. and it seems more efficient.

**Comment by jcsp** [comment]
OK - if it works then it works :-)

**Pull request merged**